contracts/treasury/liquidateWithKyber/LiquidateFeeWithKyber.sol

```diff
@@ -0,0 +1,260 @@
+  ) external onlyAdmin {
+    _setLiquidationStrategy(_strategy);
+    _setPriceOracle(_oracle);
+    _setKyberNetworkProxy(_proxy);
+  }
+
+  function manualApproveAllowancesToKyberProxy(IERC20Ext[] calldata tokens, bool isReset)
+    external onlyOperator
+  {
+    for(uint256 i = 0; i < tokens.length; i++) {
+      _safeApproveAllowance(
+        tokens[i],
+        address(kyberProxy),
+        isReset ? 0 : type(uint256).max
+      );
+    }
+  }
+
+  /**
+   * @dev Anyone can call this function to liquidate LP/normal tokens to a dest token
+   *  To save gas, should specify the list of final tokens to swap to dest token
+   *  Pass list of tradeTokens + corresponding balances before the liquidation happens
+   *    as txData, will be used to get the received amount of each token to swap
+   * @param tokens list of source tokens
+   * @param amounts amount of each source token
+   * @param types type of each token, either LP or normal token
+   * @param dest dest token to swap to
+   * @param tradeTokens list of final tokens to swap to dest token after removing liquidities
+   */
+  function liquidate(
+    IERC20Ext[] calldata tokens,
+    uint256[] calldata amounts,
+    LiquidationType[] calldata types,
+    IERC20Ext dest,
+    IERC20Ext[] calldata tradeTokens
+  ) external {
+    require(
+      tokens.length == amounts.length && amounts.length == types.length,
+      'invalid lengths'
+    );
+    // add one extra element for balance of dest token before liquidate call
+    uint256[] memory balances = new uint256[](tradeTokens.length + 1);
+    for(uint256 i = 0; i < tradeTokens.length; i++) {
+      balances[i] = getBalance(tradeTokens[i], address(this));
+    }
+    balances[tradeTokens.length] = getBalance(dest, address(this));
+    bytes memory oracleHint = abi.encode(types);
+    bytes memory txData = abi.encode(types, tradeTokens, balances);
+    liquidationStrategy.liquidate(
+      priceOracle,
+      tokens,
+      amounts,
+      payable(address(this)),
+      dest,
+      oracleHint,
+      txData
+    );
+  }
+
+  /**
+   * @dev Only accept the callback from the liquidationStrategy
+   *    remove all liquidity if needed, then swap all tradeTokens to dest token
+   */
+  function liquidationCallback(
+    address caller,
+    IERC20Ext[] calldata sources,
+    uint256[] calldata amounts,
+    address payable recipient,
+    IERC20Ext dest,
+    uint256 minReturn,
+    bytes calldata txData
+  ) external override {
+    require(msg.sender == address(liquidationStrategy), 'sender != liquidationStrategy');
+    require(caller == address(this), 'caller != this address');
+    (
+      LiquidationType[] memory types,
+      IERC20Ext[] memory tradeTokens,
+      uint256[] memory balancesBefore
+    ) = abi.decode(txData, (LiquidationType[], IERC20Ext[], uint256[]));
+
+    // last element will be the dest token balance of this contract before the callback
+    uint256 destTokenBefore = balancesBefore[balancesBefore.length - 1];
```

Comment:
gas optimization: we can remove this field since it's already part of `balancesBefore`.
`_swapWithKyber()` can drop the `destTokenBefore` field to become
`_swapWithKyber(tradeTokens, balancesBefore, dest);`

then in the totalReturn calculation
```
// last element of balancesBefore = dest token balance of this contract before liquidation callback
totalReturn = getBalance(dest, address(this)).sub(balancesBefore[balancesBefore.length - 1]);
```

Reply:
Yeah, true, previously I only save ethBalance as the last element, but later decided to just do the same for all dest tokens, so it is easier